airflow/providers/facebook/ads/hooks/ads.py

```diff
@@ -0,0 +1,152 @@
+    :type api_version: str
+
+    """
+
+    def __init__(
+        self,
+        facebook_ads_conn_id: str = "facebook_default",
+        api_version: str = "v6.0",
+    ) -> None:
+        super().__init__()
+        self.facebook_ads_conn_id = facebook_ads_conn_id
+        self.api_version = api_version
+        self.default_params = {
+            'level': 'ad',
+            'date_preset': 'yesterday',
+        }
+
+    def _get_service(self) -> FacebookAdsApi:
+        """ Returns Facebook Ads Client using a service account"""
+        config = self.facebook_ads_config
+        try:
+            return FacebookAdsApi.init(app_id=config["app_id"],
+                                       app_secret=config["app_secret"],
+                                       access_token=config["access_token"],
+                                       account_id=config["account_id"],
+                                       api_version=self.api_version)
+        except AirflowException as error:
+            message = "Missing config data:\n {error}".format(error=error)
+            self._error_handler(message)
+            raise
+
+    @cached_property
+    def facebook_ads_config(self) -> None:
+        """
+        Gets Facebook ads connection from meta db and sets
+        facebook_ads_config attribute with returned config file
+        """
+        conn = self.get_connection(self.facebook_ads_conn_id)
+        if "facebook_ads_client" not in conn.extra_dejson:
+            raise AirflowException("facebook_ads_client not found in extra field")
+        return conn.extra_dejson.get("facebook_ads_client")
+
+    def bulk_facebook_report(
+        self,
+        params: Optional[Dict[str, Any]] = None,
+        fields: Optional[List[str]] = None,
+    ) -> List[AdsInsights]:
+        """
+        Pulls data from the Facebook Ads API
+
+        :param fields: List of fields that is obtained from Facebook. Found in AdsInsights.Field class.
+            https://developers.facebook.com/docs/marketing-api/insights/parameters/v6.0
+        :type fields: List[str]
+        :param params: Parameters that determine the query for Facebook
+            https://developers.facebook.com/docs/marketing-api/insights/parameters/v6.0
+
+        :return: Facebook Ads API response, converted to Facebook Ads Row objects
+        :rtype: List[AdsInsights]
+        """
+        api = self._get_service()
+        params = params if params else self.default_params
+        ad_account = AdAccount(api.get_default_account_id(), api=api)
+        _async = ad_account.get_insights(params=params, fields=fields, is_async=True)
+        try:
+            while _async.api_get()["async_status"] != JOB_COMPLETED:
+                percent = _async.api_get()["async_percent_completion"]
+                self.log.info("%s %s completed", percent, "%")
+                time.sleep(10)
+                if _async.api_get()["async_status"] in [JOB_SKIPPED, JOB_FAILED]:
```

Comment:
This condition should be after the loop if you enter my suggestion. You will also need to check the success of the surgery.